runtime/src/accounts_db.rs

```diff
@@ -1266,6 +1266,41 @@ impl AccountsDB {
         }
     }
 
+    pub fn remove_unrooted_slot(&self, remove_slot: Slot) {
+        if self.accounts_index.read().unwrap().is_root(remove_slot) {
+            panic!("Trying to remove accounts for rooted slot {}", remove_slot);
+        }
+
+        let pubkey_sets: Vec<HashSet<Pubkey>> = self.scan_account_storage(
+            remove_slot,
+            |stored_account: &StoredAccount, _, accum: &mut HashSet<Pubkey>| {
+                accum.insert(stored_account.meta.pubkey);
+            },
+        );
+
+        // Purge this slot from the accounts index
+        {
+            let accounts_index = self.accounts_index.read().unwrap();
+            for pubkey_set in pubkey_sets {
+                for pubkey in pubkey_set {
+                    accounts_index
+                        .account_maps
+                        .get(&pubkey)
+                        .expect("account in appendvec must have entry in accounts index")
+                        .1
+                        .write()
+                        .unwrap()
+                        .retain(|(pubkey_slot, _)| *pubkey_slot != remove_slot)
```

Comment:
How about when this retain leaves no forks in the slot, doesn't seem to cleanup the index entry.

Reply:
@sakridge, do I need to keep the entry around to retain that first item (the refcount) in the tuple? If I purge here, worried the refcount might get

Reply:
This seems like it's ok to remove because i'm also removing all the storage entries, so the refcount won't get decremented. 

This seems to imply I should also decrement by the refcount by the number of times the account appears in the purged storage entries

Reply:
Actually maybe you need to check that the refcount is zero after removing the storage entries. Because the account could have a deleted version in another append_vec. @ryoqun does that sound right? Or maybe just leave it and have the regular cleanup get it later.

Reply:
Look like this is cool pr! :) I'm catching up the context... I'll comment my thinking today later.

Reply:
@sakridge ah yea makes sense

Reply:
ah, is it possible to have an empty account index list for a slot, but refcount is greater than zero?

Reply:
> is it possible to have an empty account index list for a slot, but refcount is greater than zero?

> Edit: Yeah it seems possible b/c the refcount isn't decremented until all the storage entries for the slot have a count of 0.

Well, it's NOT possible to have an empty account index list. But, it's possible the refcount is greater than 0 or the number of elements in the account index list for given pubkey.

AccountDB has rather unusual life-cycyle: It always writes newer version of account to update and _remove_ an account. It always adds something to the list. So, it always has at least 1 element in the account index list for given pubkey.

The only exception is that when it removes 0-lamport account. But it's too complex to describe here. ;) In a nutshell, when it dares to do so, it do so by purging the whole entry for given pubkey from `accounts_index.account_maps`.

> Not too sure what this means.

;) That's due to my poor commenting.. Simply put, the refcount means there is the `refcount` total number of possible references to the given pubkey from each version of account state in the entire set of `AppendVec`s in the `./accounts` dir. The fact `refcount` is `0` or not important when we're determining to remove 0-lamport tombstone version of account state for the given pubkey, while ensuring we can re-construct the index only from the account storages.

Reply:
Well, my previous comment is illustrative (just hopes..) at best. not helpful for the direction of pr.

In that regard, I think we should just hook to call `handle_reclaims()` from here somehow instead of `retain()`-ing directly here. And I rather want to avoid another bypass code path like this. `handle_reclaims()` is the battle-tested entry point to sanely and safely collect garbage regardless of rooted/non-rooted- outdated versions of account states. It's called from various places already.

Also, the assumption here is that there is no strong performance requirements here. I think this is a safe assumption because this pr is rather rare code path.

For example, 

```rust
struct AccountsDB {
    pub fn remove_unrooted_slot(&self, remove_slot: Slot) {
        ...
        let reclaims = []
        for pubkey in ... {
          ...
          account_index.clean_unrooted_entries_by_slot(pubkey, remove_slot, reclaims);
          ...
        }
    }
    self.handle_reclaims(reclaims);
    // force to process dead slots synchronously.
    self.process_dead_slots();
    assert!(....); // remove_slot is really removed? or panic!()
}

...

struct AccountsIndex {
    fn clean_unrooted_entries_by_slot(self&, pubkey, remove_slot, &mut reclaims) {
       // this is like `self.clean_rooted_entries()`
       // should be straight forward logic
    }
}
```

Also, even if we did this way or do `retain()` here, we may leave empty account index lists for pubkeys with its refcount always being zero (yes, we can ensure this). This is new data shape but it shouldn't break things. Lastly, I don't think this is major concern because we won't leave empty lists for long time.

That's because we'll eventually update the accounts (= corresponding AccountsIndex entries) later. Generally speaking, even if a bank at a particular slot is abandoned, contained TXes will be replayed later unless we starts to censor those TXs out of sudden. ;) In other words, once account updates hits the AccountsDB it's already authorized (and buffered) updates.

So that way, we're now effectively and nicely pruning never-rooted account index entries regardless replaying alternate versions of dead slots.

@sakridge @carllin this is hard topic and I'm not good at explaining. any questions are very welcome. :)

Reply:
@ryoqun, thanks for the detailed analysis! I was parsing through some of the cleanup code yesterday, but I think my understanding of how all the parts interact (especially the portion on the 0-lamport accounts) is still quite incomplete.

I think this makes sense, calling tested existing methods instead off introducing new code paths. Updated with the suggested changes!